### PR TITLE
More build options: docs, versioning

### PR DESCRIPTION
**Pull request by jcierocki**

### CODE REVIEW

File: {{ cookiecutter.repo_name }}/pyproject.toml [hunk 1]
```diff
@@ -133,13 +128,87 @@ accept-no-yields-doc = false
...
+
+
+
+{% if cookiecutter.versioning == "Python Semantic Release" %}
+exclude_dirs = ["venv"]
 # B101 disables errors for asserts in the code
 # remember to not use asserts for security and control flows
-skips = ["B101"]
\ No newline at end of file
+skips = ["B101"]
```
Comment (karllu3) [inline {{ cookiecutter.repo_name }}/pyproject.toml:145]: Exclude_dir and skips shall be independent of  "cookiecutter.versioning"

File: {{ cookiecutter.repo_name }}/pyproject.toml [hunk 2]
```diff
@@ -133,13 +128,87 @@ accept-no-yields-doc = false
...
+
+[tool.semantic_release.remote]
+name = "origin"
+type = "github"
+ignore_token_for_push = false
+insecure = false
+
+[tool.semantic_release.publish]
+
+{% endif %}
```
Comment (karllu3) [inline {{ cookiecutter.repo_name }}/pyproject.toml:212]: endif shall go to the end

File: {{ cookiecutter.repo_name }}/pyproject.toml [hunk 3]
```diff
@@ -133,13 +128,87 @@ accept-no-yields-doc = false
...
-min-similarity-lines=10
+min-similarity-lines = 10
 
 [tool.bandit]
-exclude_dirs = ["venv",]
+
+
+
+
+{% if cookiecutter.versioning == "Python Semantic Release" %}
```
Comment (karllu3) [inline {{ cookiecutter.repo_name }}/pyproject.toml:141]: To be honest I would leave this without if it shall not break anything.
If You want go with this if than {% raw %} is missing to escape e.g block_start_string = {%

File: {{ cookiecutter.repo_name }}/pyproject.toml (hunk 2 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/pyproject.toml]: typo, fixed

File: {{ cookiecutter.repo_name }}/pyproject.toml (hunk 1 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/pyproject.toml:140]: typo, fixed

File: cookiecutter.json [hunk 4]
```diff
@@ -1,21 +1,14 @@
...
-    ]
-  }
\ No newline at end of file
+  "client_name": "ds",
+  "project_name": "default",
+  "__project_name_slug": "{{cookiecutter.project_name | slugify}}",
+  "__client_name_slug": "{{cookiecutter.client_name | slugify}}",
+  "repo_name": "{{cookiecutter.__client_name_slug}}-{{cookiecutter.__project_name_slug}}",
+  "ci": ["GitLab", "Github", "None"],
+  "jupytext": ["No", "Yes"],
```
Comment (karllu3) [inline cookiecutter.json:8]: If You don't mind I would change order of options versioning 
 "Python Semantic Release",  "Bumpversion", "None",
 for ci:
 "Github", "Gitlab", "None"

File: cookiecutter.json [hunk 5]
```diff
@@ -1,21 +1,14 @@
...
+  "docs": ["Sphinx", "No docs"],
+  "versioning": ["Bumpversion", "Python Semantic Release", "None"],
+  "python_package_name": "{{ cookiecutter.__project_name_slug.replace('-', '_') }}",
+  "__package_name": "{{ cookiecutter.python_package_name }}",
+  "_copy_without_render": [".github/workflows/*.yml"]
```
Comment (karllu3) [inline cookiecutter.json:13]: I would also add  ".github/workflows/*.yaml"

File: cookiecutter.json (hunk 4 above)
Comment (jcierocki) [reply cookiecutter.json:8]: Ok

File: cookiecutter.json (hunk 5 above)
Comment (jcierocki) [reply cookiecutter.json:13]: Ok

File: .github/ISSUE_TEMPLATE/01_feature_request.yml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+name: 🚀 Feature Request
```
Comment (mateusz-wosinski-ds) [inline .github/ISSUE_TEMPLATE/01_feature_request.yml:1]: skąd to tutaj?

Comment (jcierocki) [reply .github/ISSUE_TEMPLATE/01_feature_request.yml:1]: Dobre pytanie. Wyczyszę

File: .github/ISSUE_TEMPLATE/02_bug_report.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+name: 🐞 Bug Report
```
Comment (mateusz-wosinski-ds) [inline .github/ISSUE_TEMPLATE/02_bug_report.yml:1]: as above

File: cookiecutter.json [hunk 8]
```diff
@@ -1,21 +1,14 @@
...
\ No newline at end of file
+  "client_name": "ds",
+  "project_name": "default",
+  "__project_name_slug": "{{cookiecutter.project_name | slugify}}",
+  "__client_name_slug": "{{cookiecutter.client_name | slugify}}",
+  "repo_name": "{{cookiecutter.__client_name_slug}}-{{cookiecutter.__project_name_slug}}",
+  "ci": ["Github", "GitLab", "None"],
+  "jupytext": ["No", "Yes"],
+  "docs": ["Sphinx", "No docs"],
+  "versioning": ["Python Semantic Release", "Bumpversion", "None"],
```
Comment (mateusz-wosinski-ds) [inline cookiecutter.json:10]: do we support python semantic already? :thinking:

File: hooks/post_gen_project.py [hunk 9]
```diff
@@ -4,33 +4,59 @@
 
 print("Running post generation...")
 
-ci = "{{ cookiecutter.ci }}"
+files_to_be_removed = []
```
Comment (mateusz-wosinski-ds) [inline hooks/post_gen_project.py:7]: why isn't it written in uppercase? It is a global variable as well

File: hooks/pre_gen_project.py [hunk 10]
```diff
@@ -17,4 +17,11 @@
     print(f"ERROR: '{package_name}' is not a valid Python module name.")
     sys.exit(1)
 
+# TODO temporary, until GitLab CI support for Semantic Release will be added
+ci_type = '{{ cookiecutter.ci }}'
+versioning_approach = '{{ cookiecutter.versioning }}'
+if ci_type == 'GitLab' and versioning_approach == "Python Semantic Release":
+    print(f"ERROR: 'Python Semantic Release' is currently not supported for projects with GitLab CI.")
+    sys.exit(1)
```
Comment (mateusz-wosinski-ds) [inline hooks/pre_gen_project.py:25]: I would suggest adding Python Semantic as the option for versioning approach only after adding methods for supporting it instead of printing some errors

File: {{ cookiecutter.repo_name }}/.github/workflows/check_future_version.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Python Semantic Release
+        uses: python-semantic-release/python-semantic-release@master
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          root_options: "-v --noop"
```
Comment (mateusz-wosinski-ds) [inline {{ cookiecutter.repo_name }}/.github/workflows/check_future_version.yml:22]: I would not mix many changes in a single PR - this one should focus on buliding options, not adding some Python Semantic-related changes

File: {{ cookiecutter.repo_name }}/.github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
+name: Semantic Release
```
Comment (mateusz-wosinski-ds) [inline {{ cookiecutter.repo_name }}/.github/workflows/release.yml:1]: as above

File: {{ cookiecutter.repo_name }}/.gitlab-ci.yml [hunk 13]
```diff
@@ -236,6 +235,7 @@ tests:
   rules:
     - if: $CI_COMMIT_BRANCH == $CI_DEFAULT_BRANCH
 
+{% if cookiecutter.docs != "No docs" %}
```
Comment (mateusz-wosinski-ds) [inline {{ cookiecutter.repo_name }}/.gitlab-ci.yml:238]: I would rather move it to `rules` part of the `pages` stage

File: {{ cookiecutter.repo_name }}/README.md [hunk 14]
```diff
@@ -118,11 +121,28 @@ $ ./bump_version.sh minor
...
 # to see what is going to change run:
-$ ./bump_version.sh --dry-run major
+$ ./bump_versemantic-release version --patch --printsion.sh --dry-run major
 ```
 Script updates **VERSION** file and setup.cfg automatically uses that version.
 
 You can configure it to update version string in other files as well - please check out the bump2version configuration file.
+{% endif -%}
+
+{%- if cookiecutter.versioning == "Python Semantic Release" %}
```
Comment (mateusz-wosinski-ds) [inline {{ cookiecutter.repo_name }}/README.md:131]: as above, should be added in a separate PR concerning Python Semantic only

File: {{ cookiecutter.repo_name }}/pyproject.toml [hunk 15]
```diff
@@ -133,13 +128,84 @@ accept-no-yields-doc = false
...
 [tool.bandit]
-exclude_dirs = ["venv",]
+exclude_dirs = ["venv"]
 # B101 disables errors for asserts in the code
 # remember to not use asserts for security and control flows
-skips = ["B101"]
\ No newline at end of file
+skips = ["B101"]
+
+{% if cookiecutter.versioning == "Python Semantic Release" %}
```
Comment (mateusz-wosinski-ds) [inline {{ cookiecutter.repo_name }}/pyproject.toml:142]: as above

File: {{ cookiecutter.repo_name }}/src/{{ cookiecutter.__package_name }}/VERSION [hunk 16]
```diff
@@ -1 +0,0 @@
-0.0.1-dev
```
Comment (mateusz-wosinski-ds) [inline {{ cookiecutter.repo_name }}/src/{{ cookiecutter.__package_name }}/VERSION:1]: Im not sure it would not break anything - there was some reason for adding this file :thinking:

**Review by mateusz-wosinski-ds — CHANGES_REQUESTED** [review]
please don't mix changes concerning 2 features in one Pull Request :pray: CI tests fail in the current implementation, so pls test the updated version locally before pushing ;)

File: cookiecutter.json (hunk 8 above)
Comment (jcierocki) [reply cookiecutter.json:10]: Tak, ale tylko dla GitHub CI/CD

File: hooks/pre_gen_project.py (hunk 10 above)
Comment (jcierocki) [reply hooks/pre_gen_project.py:25]: Good idea!

File: {{ cookiecutter.repo_name }}/.github/workflows/check_future_version.yml (hunk 11 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/.github/workflows/check_future_version.yml:22]: The thing is this stuff is ready and also it seems normal that while making versioning optional we added another option for it.

File: {{ cookiecutter.repo_name }}/src/{{ cookiecutter.__package_name }}/VERSION (hunk 16 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/src/{{ cookiecutter.__package_name }}/VERSION:1]: We can keep it, but according to @mhordynski we should switch to `__version__.py` only.

File: .github/ISSUE_TEMPLATE/01_feature_request.yml (hunk 6 above)
Comment (jcierocki) [reply .github/ISSUE_TEMPLATE/01_feature_request.yml:1]: Właśnie sprawdziłem i dodał to @mhordynski. Nie wiem na ile mogę to usuwać

File: hooks/post_gen_project.py (hunk 9 above)
Comment (jcierocki) [reply hooks/post_gen_project.py:7]: But it's not a typical constant because it is getting mutated later down the track. Uppercase is for constants (hyperparameters) not for any global variables, at least for me.

File: {{ cookiecutter.repo_name }}/.gitlab-ci.yml (hunk 13 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/.gitlab-ci.yml:238]: Good idea!

File: {{ cookiecutter.repo_name }}/.github/workflows/check_future_version.yml (hunk 11 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/.github/workflows/check_future_version.yml]: Done. All _Python Semantic Release_ stuff is not in #24

File: {{ cookiecutter.repo_name }}/.gitlab-ci.yml (hunk 13 above)
Comment (jcierocki) [reply {{ cookiecutter.repo_name }}/.gitlab-ci.yml]: Done.

File: hooks/pre_gen_project.py (hunk 10 above)
Comment (jcierocki) [reply hooks/pre_gen_project.py]: No longer applies after _Python Semantic Release_ got scraped from this PR.

**Review by Sahcim — APPROVED** [review]
I merged this branch with main ( to remove files from changes ). Fixed few typos since @jcierocki is on vacation. LGTM we can merge it.
